pkg/api/routes.go

```diff
@@ -488,11 +516,130 @@ func (rh *RouteHandler) UpdateManifest(response http.ResponseWriter, request *ht
 		return
 	}
 
+	if rh.c.RepoDB != nil {
+		// check is image is a signature
+		isSignature, signatureType, signedManifestDigest, err := storage.CheckIsImageSignature(name, body, reference,
+			rh.c.StoreController)
+		if err != nil {
+			if errors.Is(err, zerr.ErrOrphanSignature) {
+				rh.c.Log.Warn().Err(err).Msg("image has signature format but it doesn't sign any image")
+
+				response.Header().Set("Location", fmt.Sprintf("/v2/%s/manifests/%s", name, digest))
+				response.Header().Set(constants.DistContentDigestKey, digest.String())
+				response.WriteHeader(http.StatusCreated)
+
+				return
+			}
+
+			rh.c.Log.Error().Err(err).Msg("can't check if image is a signature or not")
+
+			if err = imgStore.DeleteImageManifest(name, reference); err != nil {
+				rh.c.Log.Error().Err(err).Msgf("couldn't remove image manifest %s in repo %s", reference, name)
+			}
+
+			response.WriteHeader(http.StatusInternalServerError)
+
+			return
+		}
+
+		metadataSuccessfullySet := true
+
+		if isSignature {
+			err := rh.c.RepoDB.AddManifestSignature(signedManifestDigest, repodb.SignatureMetadata{
+				SignatureType:   signatureType,
+				SignatureDigest: digest,
+			})
+			if err != nil {
+				rh.c.Log.Error().Err(err).Msg("repodb: error while putting repo meta")
+				metadataSuccessfullySet = false
+			}
+		} else {
+			imageMetadata, err := newManifestMeta(name, body, rh.c.StoreController)
+			if err == nil {
+				err := rh.c.RepoDB.SetManifestMeta(digest, imageMetadata)
+				if err != nil {
+					rh.c.Log.Error().Err(err).Msg("repodb: error while putting image meta")
+					metadataSuccessfullySet = false
+				} else {
+					// If SetManifestMeta is successful and SetRepoTag is not, the data inserted by SetManifestMeta
+					// will be garbage collected later
+					// Q: There will be a problem if we write a manifest without a tag
+					// Q: When will we write a manifest where the reference will be a digest?
+					err = rh.c.RepoDB.SetRepoTag(name, reference, digest)
+					if err != nil {
+						rh.c.Log.Error().Err(err).Msg("repodb: error while putting repo meta")
+						metadataSuccessfullySet = false
+					}
+				}
+			} else {
+				metadataSuccessfullySet = false
+			}
+		}
+
+		if !metadataSuccessfullySet {
+			rh.c.Log.Info().Msgf("uploding image meta was unsuccessful for tag %s in repo %s", reference, name)
+
+			if err = imgStore.DeleteImageManifest(name, reference); err != nil {
+				rh.c.Log.Error().Err(err).Msgf("couldn't remove image manifest %s in repo %s", reference, name)
+			}
+
+			response.WriteHeader(http.StatusInternalServerError)
+
+			return
+		}
+	}
+
 	response.Header().Set("Location", fmt.Sprintf("/v2/%s/manifests/%s", name, digest))
 	response.Header().Set(constants.DistContentDigestKey, digest.String())
 	response.WriteHeader(http.StatusCreated)
 }
 
+func newManifestMeta(repoName string, manifestBlob []byte,
```

Comment:
Add comments on what this method does, and why it is needed etc ... mainly for our sake